doc/man/man1/ldapcompare.1
.TH LDAPCOMPARE 1 "RELEASEDATE" "OpenLDAP LDVERSION"
.\" $OpenLDAP$
.\" Copyright 1998-2026 The OpenLDAP Foundation All Rights Reserved.
.\" Copying restrictions apply.  See COPYRIGHT/LICENSE.
.SH NAME
ldapcompare \- LDAP compare tool
.SH SYNOPSIS
.B ldapcompare
[\c
.BR \-V [ V ]]
[\c
.BI \-d \ debuglevel\fR]
[\c
.BR \-n ]
[\c
.BR \-v ]
[\c
.BR \-z ]
[\c
.BR \-M [ M ]]
[\c
.BR \-x ]
[\c
.BI \-D \ binddn\fR]
[\c
.BR \-W ]
[\c
.BI \-w \ passwd\fR]
[\c
.BI \-y \ passwdfile\fR]
[\c
.BI \-H \ ldapuri\fR]
[\c
.BR \-P \ { 2 \||\| 3 }]
[\c
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]]
[\c
.BI \-o \ opt \fR[= optparam \fR]]
[\c
.BI \-O \ security-properties\fR]
[\c
.BR \-I ]
[\c
.BR \-Q ]
[\c
.BR \-N ]
[\c
.BI \-U \ authcid\fR]
[\c
.BI \-R \ realm\fR]
[\c
.BI \-X \ authzid\fR]
[\c
.BI \-Y \ mech\fR]
[\c
.BR \-Z [ Z ]]
.IR DN
{\c
.BI attr: value
|
.BI attr:: b64value\fR}
.SH DESCRIPTION
.I ldapcompare
is a shell-accessible interface to the
.BR ldap_compare_ext (3)
library call.
.LP
.B ldapcompare
opens a connection to an LDAP server, binds, and performs a compare
using specified parameters.   The \fIDN\fP should be a distinguished
name in the directory.  \fIAttr\fP should be a known attribute.  If
followed by one colon, the assertion \fIvalue\fP should be provided
as a string.  If followed by two colons, the base64 encoding of the
value is provided.  The result code of the compare is provided as
the exit code and, unless ran with \fB\-z\fP, the program prints
TRUE, FALSE, or UNDEFINED on standard output.
.LP
.SH OPTIONS
.TP
.BR \-V [ V ]
Print version info.
If \fB\-VV\fP is given, only the version information is printed.
.TP
.BI \-d \ debuglevel
Set the LDAP debugging level to \fIdebuglevel\fP.
.B ldapcompare
must be compiled with LDAP_DEBUG defined for this option to have any effect.
.TP
.B \-n
Show what would be done, but don't actually perform the compare.  Useful for
debugging in conjunction with \fB\-v\fP.
.TP
.B \-v
Run in verbose mode, with many diagnostics written to standard output.
.TP
.B \-z
Run in quiet mode, no output is written.  You must check the return
status.  Useful in shell scripts.
.TP
.BR \-M [ M ]
Enable manage DSA IT control.
.B \-MM
makes control critical.
.TP
.B \-x 
Use simple authentication instead of SASL.
.TP
.BI \-D \ binddn
Use the Distinguished Name \fIbinddn\fP to bind to the LDAP directory.
For SASL binds, the server is expected to ignore this value.
.TP
.B \-W
Prompt for simple authentication.
This is used instead of specifying the password on the command line.
.TP
.BI \-w \ passwd
Use \fIpasswd\fP as the password for simple authentication.
.TP
.BI \-y \ passwdfile
Use complete contents of \fIpasswdfile\fP as the password for
simple authentication.
Note that \fIcomplete\fP means that any leading or trailing whitespaces,
including newlines, will be considered part of the password and,
unlike other software, they will not be stripped.
As a consequence, passwords stored in files by commands like
.BR echo (1)
will not behave as expected, since
.BR echo (1)
by default appends a trailing newline to the echoed string.
The recommended portable way to store a cleartext password in a file
for use with this option is to use
.BR slappasswd (8)
with \fI{CLEARTEXT}\fP as hash and the option \fB\-n\fP.
.TP
.BI \-H \ ldapuri
Specify URI(s) referring to the ldap server(s); only the protocol/host/port
fields are allowed; a list of URI, separated by whitespace or commas
is expected.
.TP
.BR \-P \ { 2 \||\| 3 }
Specify the LDAP protocol version to use.
.TP
.BR \-e \ [ ! ] \fIext\fP [ =\fIextparam\fP ]
.TP
.BR \-E \ [ ! ] \fIext\fP [ =\fIextparam\fP ]

Specify general extensions with \fB\-e\fP and compare extensions with \fB\-E\fP.
\'\fB!\fP\' indicates criticality.

General extensions:
.nf
  [!]assert=<filter>    (an RFC 4515 Filter)
  !authzid=<authzid>    ("dn:<dn>" or "u:<user>")
  [!]bauthzid           (RFC 3829 authzid control)
  [!]chaining[=<resolve>[/<cont>]]
  [!]manageDSAit
  [!]noop
  ppolicy
  [!]postread[=<attrs>] (a comma-separated attribute list)
  [!]preread[=<attrs>]  (a comma-separated attribute list)
  [!]relax
  sessiontracking[=<username>]
  abandon,cancel,ignore (SIGINT sends abandon/cancel,
  or ignores response; if critical, doesn't wait for SIGINT.
  not really controls)
.fi

Compare extensions:
.nf
  !dontUseCopy
.fi
.TP
.BI \-o \ opt \fR[= optparam \fR]

Specify any
.BR ldap.conf (5)
option or one of the following:
.nf
  nettimeout=<timeout>  (in seconds, or "none" or "max")
  ldif_wrap=<width>     (in columns, or "no" for no wrapping)
.fi

.TP
.BI \-O \ security-properties
Specify SASL security properties.
.TP
.B \-I
Enable SASL Interactive mode.  Always prompt.  Default is to prompt
only as needed.
.TP
.B \-Q
Enable SASL Quiet mode.  Never prompt.
.TP
.B \-N
Do not use reverse DNS to canonicalize SASL host name.
.TP
.BI \-U \ authcid
Specify the authentication ID for SASL bind. The form of the ID
depends on the actual SASL mechanism used.
.TP
.BI \-R \ realm
Specify the realm of authentication ID for SASL bind. The form of the realm
depends on the actual SASL mechanism used.
.TP
.BI \-X \ authzid
Specify the requested authorization ID for SASL bind.
.I authzid
must be one of the following formats:
.BI dn: "<distinguished name>"
or
.BI u: <username>
.TP
.BI \-Y \ mech
Specify the SASL mechanism to be used for authentication. If it's not
specified, the program will choose the best mechanism the server knows.
.TP
.BR \-Z [ Z ]
Issue StartTLS (Transport Layer Security) extended operation. If you use
\fB\-ZZ\fP, the command will require the operation to be successful.
.SH EXAMPLES
.nf
    ldapcompare "uid=babs,dc=example,dc=com"  sn:Jensen
    ldapcompare "uid=babs,dc=example,dc=com"  sn::SmVuc2Vu
.fi
are all equivalent.
.SH LIMITATIONS
Requiring the value be passed on the command line is limiting
and introduces some security concerns.  The command should support
a mechanism to specify the location (file name or URL) to read
the value from.
.SH "SEE ALSO"
.BR ldap.conf (5),
.BR ldif (5),
.BR ldap (3),
.BR ldap_compare_ext (3)
.SH AUTHOR
The OpenLDAP Project <http://www.openldap.org/>
.SH ACKNOWLEDGEMENTS
.so ../Project


CHANGES
OpenLDAP 2.6 Change Log

OpenLDAP 2.6.13 Engineering

OpenLDAP 2.6.12 Release (2026/01/29)
	Fixed libldap to reject empty types in LDIF (ITS#10429)
	Fixed libldap to not scroll past nul bytes (ITS#10430)
	Fixed libldap to enforce stop when encountering nul-leading line (ITS#10431)

OpenLDAP 2.6.11 Release (2026/01/28)
	Fixed slapd to use fresh timestamp for lastbind (ITS#10379)
	Fixed slapd delta-syncrepl to always use logDB rootdn (ITS#10360)
	Fixed slapd reverse lookup of proxied IPv6 addresses (ITS#10387)
	Fixed slapd logging buffer overflow (ITS#10410)
	Fixed slapd-ldap response when invalid secprops is configured (ITS#10392)
	Fixed slapd-mdb error when deleting last child of a branch (ITS#10304)
	Fixed slapd-mdb check for pool pause in search (ITS#10191)
	Fixed slapo-memberof clash with refint on subtree rename (ITS#10398)
	Fixed slapo-syncprov use correct rootDN for accesslog replay (ITS#10385)
	Minor Cleanup
		ITS#7901
		ITS#10329
		ITS#10335
		ITS#10339
		ITS#10343
		ITS#10344
		ITS#10345
		ITS#10347
		ITS#10348
		ITS#10349
		ITS#10353
		ITS#10358
		ITS#10359
		ITS#10366
		ITS#10367
		ITS#10369
		ITS#10370
		ITS#10371
		ITS#10372
		ITS#10374
		ITS#10375
		ITS#10376
		ITS#10377
		ITS#10379
		ITS#10380
		ITS#10381
		ITS#10384
		ITS#10388
		ITS#10390
		ITS#10391
		ITS#10400
		ITS#10401
		ITS#10408

OpenLDAP 2.6.10 Release (2025/05/22)
	Added slapd microsecond timestamp format for local logging (ITS#10140)
	Fixed libldap ldap_result behavior with LDAP_MSG_RECEIVED (ITS#10229)
	Fixed lloadd handling of starttls critical (ITS#10323)
	Fixed slapd syncrepl when used with slapo-rwm (ITS#10290)
	Fixed slapd regression with certain searches (ITS#10307)
	Fixed slapo-autoca olcAutoCAserverClass object (ITS#10288)
	Fixed slapo-pcache caching behaviors (ITS#10270)
	Minor Cleanup
		ITS#7080
		ITS#7249
		ITS#9934
		ITS#10020
		ITS#10168
		ITS#10226
		ITS#10279
		ITS#10299
		ITS#10302
		ITS#10309
		ITS#10312
		ITS#10320
		ITS#10325
		ITS#10327
		ITS#10328
		ITS#10331
		ITS#10336

OpenLDAP 2.6.9 Release (2024/11/26)
	Fixed libldap TLS connection timeout handling (ITS#8047)
	Fixed libldap GnuTLS incompatible pointer type (ITS#10253)
	Fixed libldap OpenSSL set_ciphersuite error handling (ITS#10223)
	Fixed libldap to check for OpenSSL EVP_Digest* failure (ITS#10224)
	Fixed slapd cn=config disallowed modification of cn=schema (ITS#10256)
	Fixed slapd syncrepl assert during refresh at shutdown (ITS#10232)
	Fixed slapd syncrepl retry state during refreshDone (ITS#10234)
	Fixed slapd-ldap use of multi-precision add for op counters (ITS#10237)
	Fixed slapd-mdb idl intersection (ITS#10233)
	Fixed slapd-wt idl intersection (ITS#10233)
	Fixed slapo-memberof to omit dynamic values (ITS#10230)
	Fixed slapo-nestgroup leak in nestgroup_memberFilter (ITS#10249)
	Fixed slapo-translucent regression with subordinate databases (ITS#10248)
	Fixed slapo-translucent regression when requesting attributes (ITS#10272)
	Fixed slappw-argon2 defaults to be more secure (ITS#9827)
	Minor Cleanup
		ITS#10155
		ITS#10218
		ITS#10219
		ITS#10227
		ITS#10231
		ITS#10235
		ITS#10263
		ITS#10264

OpenLDAP 2.6.8 Release (2024/05/21)
	Fixed libldap exit handling with OpenSSL3 again (ITS#9952)
	Fixed libldap OpenSSL channel binding digest (ITS#10216)
	Fixed slapd handling of large uid/gids peercred auth (ITS#10211)
	Fixed slapd-asyncmeta/meta target structure allocations (ITS#10197)
	Fixed slapd-meta with dynlist (ITS#10164)
	Fixed slapd-meta binds when proxying internal op (ITS#10165)
	Added slapo-nestgroup overlay (ITS#10161)
	Added slapo-memberof 'addcheck' option (ITS#10167)
	Fixed slapo-accesslog startup initialization (ITS#10170)
	Fixed slapo-constraint double free on invalid attr (ITS#10204)
	Fixed slapo-dynlist with abandoned operations (ITS#10044)
	Build
		Fixed build with gcc14.x (ITS#10166)
		Fixed back-perl with clang15 (ITS#10177)
		Fixed to reduce systemd dependencies (ITS#10214)
	Contrib
		Added slapo-alias contrib module (ITS#10104, ITS#10182)
		Fixed slapo-autogroup to work with slapo-dynlist (ITS#10185)
		Fixed smbk5pwd implicit function declaration (ITS#10206)
	Documentation
		Fixed slapo-memberof exattr requirements (ITS#7400)
		Fixed slapo-memberof is no longer deprecated (ITS#7400)
	Minor Cleanup
		ITS#9921
		ITS#10103
		ITS#10171
		ITS#10172
		ITS#10173
		ITS#10179
		ITS#10183
		ITS#10186
		ITS#10188
		ITS#10193
		ITS#10209

OpenLDAP 2.6.7 Release (2024/01/29)
	Added slapo-dynlist option to disable filter support (ITS#10025)
	Fixed liblber missing newline on long msg (ITS#10105)
	Fixed libldap exit handling with OpenSSL3 (ITS#9952)
	Fixed libldap with TLS and multiple ldap URIs (ITS#10101)
	Fixed libldap OpenSSL cipher suite handling (ITS#10094)
	Fixed libldap OpenSSL 3.0 and Diffie-Hellman param files (ITS#10124)
	Fixed libldap timestamps on Windows (ITS#10100)
	Fixed lloadd to work when resolv.conf is missing (ITS#10070)
	Fixed lloadd handling of closing connection (ITS#10083)
	Fixed lloadd tiers to be correctly linked on startup (ITS#10142)
	Fixed slapd to honour disclose in matchedDN handling (ITS#10139)
	Fixed slapd handling of regex testing in ACLs (ITS#10089)
	Fixed slapd sync replication with glued database (ITS#10080)
	Fixed slapd local logging on Windows (ITS#10092)
	Fixed slapd-asyncmeta when remote suffix is empty (ITS#10076)
	Fixed slapo-dynlist so it can't be global (ITS#10091)
	Build
		Fixed lloadd type mismatches (ITS#10074)
		Fixed builds for Windows (ITS#10117)
		Fixed build with clang16 (ITS#10123)
	Documentation
		Fixed slapo-homedir(5) attribute name for olcHomedirArchivePath (ITS#10057)
	Minor Cleanup
		ITS#10059
		ITS#10068
		ITS#10098
		ITS#10109
		ITS#10110
		ITS#10129
		ITS#10130
		ITS#10135
		ITS#10143
		ITS#10144
		ITS#10145
		ITS#10153

OpenLDAP 2.6.6 Release (2023/07/31)
	Fixed slapd cn=config incorrect handling of paused (ITS#10045)
	Fixed slapd-meta to account for MOD ops being optional (ITS#10067)
	Fixed slapd-asyncmeta to account for MOD ops being optional (ITS#10067)


clients/tools/ldapcompare.c
/* ldapcompare.c -- LDAP compare tool */
/* $OpenLDAP$ */
/* This work is part of OpenLDAP Software <http://www.openldap.org/>.
 *
 * Copyright 1998-2026 The OpenLDAP Foundation.
 * Portions Copyright 1998-2003 Kurt D. Zeilenga.
 * Portions Copyright 1998-2001 Net Boolean Incorporated.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted only as authorized by the OpenLDAP
 * Public License.
 *
 * A copy of this license is available in the file LICENSE in the
 * top-level directory of the distribution or, alternatively, at
 * <http://www.OpenLDAP.org/license.html>.
 */
/* Portions Copyright (c) 1992-1996 Regents of the University of Michigan.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms are permitted
 * provided that this notice is preserved and that due credit is given
 * to the University of Michigan at Ann Arbor.  The name of the
 * University may not be used to endorse or promote products derived
 * from this software without specific prior written permission.  This
 * software is provided ``as is'' without express or implied warranty.
 */
/* Portions Copyright 2002, F5 Networks, Inc, All rights reserved.
 * This software is not subject to any license of F5 Networks.
 * This is free software; you can redistribute and use it
 * under the same terms as OpenLDAP itself.
 */
/* ACKNOWLEDGEMENTS:
 * This work was originally developed by Jeff Costlow (F5 Networks)
 * based, in part, on existing LDAP tools and adapted for inclusion
 * into OpenLDAP Software by Kurt D. Zeilenga.
 */

#include "portable.h"

#include <stdio.h>

#include <ac/stdlib.h>

#include <ac/ctype.h>
#include <ac/string.h>
#include <ac/unistd.h>
#include <ac/errno.h>
#include <ac/socket.h>
#include <ac/time.h>
#include <sys/stat.h>

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_IO_H
#include <io.h>
#endif

#include <ldap.h>

#include "lutil.h"
#include "lutil_ldap.h"
#include "ldap_defaults.h"

#include "common.h"


static int quiet = 0;


void
usage( void )
{
	fprintf( stderr, _("usage: %s [options] DN <attr:value|attr::b64value>\n"), prog);
	fprintf( stderr, _("where:\n"));
	fprintf( stderr, _("  DN\tDistinguished Name\n"));
	fprintf( stderr, _("  attr\tassertion attribute\n"));
	fprintf( stderr, _("  value\tassertion value\n"));
	fprintf( stderr, _("  b64value\tbase64 encoding of assertion value\n"));

	fprintf( stderr, _("Compare options:\n"));
	fprintf( stderr, _("  -E [!]<ext>[=<extparam>] compare extensions (! indicates criticality)\n"));
	fprintf( stderr, _("             !dontUseCopy                (Don't Use Copy)\n"));
	fprintf( stderr, _("  -M         enable Manage DSA IT control (-MM to make critical)\n"));
	fprintf( stderr, _("  -P version protocol version (default: 3)\n"));
	fprintf( stderr, _("  -z         Quiet mode,"
		" don't print anything, use return values\n"));
	tool_common_usage();
	exit( EXIT_FAILURE );
}

static int docompare LDAP_P((
	LDAP *ld,
	char *dn,
	char *attr,
	struct berval *bvalue,
	int quiet,
	LDAPControl **sctrls,
	LDAPControl **cctrls));


const char options[] = "z"
	"Cd:D:e:H:IMnNO:o:P:QR:U:vVw:WxX:y:Y:Z";

#ifdef LDAP_CONTROL_DONTUSECOPY
int dontUseCopy = 0;
#endif

int
handle_private_option( int i )
{
	char	*control, *cvalue;
	int		crit;

	switch ( i ) {
	case 'E': /* compare extensions */
		if( protocol == LDAP_VERSION2 ) {
			fprintf( stderr, _("%s: -E incompatible with LDAPv%d\n"),
				prog, protocol );
			exit( EXIT_FAILURE );
		}

		/* should be extended to support comma separated list of
		 *	[!]key[=value] parameters, e.g.  -E !foo,bar=567
		 */

		crit = 0;
		cvalue = NULL;
		if( optarg[0] == '!' ) {
			crit = 1;
			optarg++;
		}

		control = optarg;
		if ( (cvalue = strchr( control, '=' )) != NULL ) {
			*cvalue++ = '\0';
		}

#ifdef LDAP_CONTROL_DONTUSECOPY
		if ( strcasecmp( control, "dontUseCopy" ) == 0 ) {
			if( dontUseCopy ) {
				fprintf( stderr,
					_("dontUseCopy control previously specified\n"));
				exit( EXIT_FAILURE );
			}
			if( cvalue != NULL ) {
				fprintf( stderr,
					_("dontUseCopy: no control value expected\n") );
				usage();
			}
			if( !crit ) {
				fprintf( stderr,
					_("dontUseCopy: critical flag required\n") );
				usage();
			}

			dontUseCopy = 1 + crit;
		} else
#endif
		{
			fprintf( stderr,
				_("Invalid compare extension name: %s\n"), control );
			usage();
		}
		break;

	case 'z':
		quiet = 1;
		break;

	default:
		return 0;
	}
	return 1;
}


int
main( int argc, char **argv )
{
	char		*compdn = NULL, *attrs = NULL;
	char		*sep;
	int		rc;
	LDAP		*ld = NULL;
	struct berval	bvalue = { 0, NULL };
	int		i = 0; 
	LDAPControl	c[1];


	tool_init( TOOL_COMPARE );
	prog = lutil_progname( "ldapcompare", argc, argv );

	tool_args( argc, argv );

	if ( argc - optind != 2 ) {
		usage();
	}

	compdn = argv[optind++];
	attrs = argv[optind++];

	/* user passed in only 2 args, the last one better be in
	 * the form attr:value or attr::b64value
	 */
	sep = strchr(attrs, ':');
	if (!sep) {
		usage();
	}

	*sep++='\0';
	if ( *sep != ':' ) {
		bvalue.bv_val = strdup( sep );
		bvalue.bv_len = strlen( bvalue.bv_val );

	} else {
		/* it's base64 encoded. */
		bvalue.bv_val = malloc( strlen( &sep[1] ));
		bvalue.bv_len = lutil_b64_pton( &sep[1],
			(unsigned char *) bvalue.bv_val, strlen( &sep[1] ));

		if (bvalue.bv_len == (ber_len_t)-1) {
			fprintf(stderr, _("base64 decode error\n"));
			exit(-1);
		}
	}

	ld = tool_conn_setup( 0, 0 );

	tool_bind( ld );

	if ( 0
#ifdef LDAP_CONTROL_DONTUSECOPY
		|| dontUseCopy
#endif
		)
	{
#ifdef LDAP_CONTROL_DONTUSECOPY
		if ( dontUseCopy ) {  
			c[i].ldctl_oid = LDAP_CONTROL_DONTUSECOPY;
			c[i].ldctl_value.bv_val = NULL;
			c[i].ldctl_value.bv_len = 0;
			c[i].ldctl_iscritical = dontUseCopy > 1;
			i++;    
		}
#endif
	}

	tool_server_controls( ld, c, i );

	if ( verbose ) {
		fprintf( stderr, _("DN:%s, attr:%s, value:%s\n"),
			compdn, attrs, sep );
	}

	rc = docompare( ld, compdn, attrs, &bvalue, quiet, NULL, NULL );

	free( bvalue.bv_val );

	tool_exit( ld, rc );
}


static int docompare(
	LDAP *ld,
	char *dn,
	char *attr,
	struct berval *bvalue,
	int quiet,
	LDAPControl **sctrls,
	LDAPControl **cctrls )
{
	int		rc, msgid, code;
	LDAPMessage	*res;
	char		*matcheddn;
	char		*text;
	char		**refs;
	LDAPControl **ctrls = NULL;

	if ( dont ) {
		return LDAP_SUCCESS;
	}

	rc = ldap_compare_ext( ld, dn, attr, bvalue,
		sctrls, cctrls, &msgid );
	if ( rc == -1 ) {
		return( rc );
	}

	for ( ; ; ) {
		struct timeval	tv;

		tv.tv_sec = 0;
		tv.tv_usec = 100000;

		if ( tool_check_abandon( ld, msgid ) ) {
			return LDAP_CANCELLED;
		}

		rc = ldap_result( ld, LDAP_RES_ANY, LDAP_MSG_ALL, &tv, &res );
		if ( rc < 0 ) {
			rc = tool_perror2( ld, "ldap_result" );
			return rc;
		}

		if ( rc != 0 ) {
			break;
		}
	}

	rc = ldap_parse_result( ld, res, &code, &matcheddn, &text, &refs, &ctrls, 1 );

	if( rc != LDAP_SUCCESS ) {
		fprintf( stderr, "%s: ldap_parse_result: %s (%d)\n",
			prog, ldap_err2string( rc ), rc );
		return rc;
	}

	if ( !quiet && ( verbose || ( code != LDAP_SUCCESS && code != LDAP_COMPARE_TRUE && code != LDAP_COMPARE_FALSE )||
		(matcheddn && *matcheddn) || (text && *text) || (refs && *refs) ) )
	{
		printf( _("Compare Result: %s (%d)\n"),
			ldap_err2string( code ), code );

		if( text && *text ) {
			printf( _("Additional info: %s\n"), text );
		}

		if( matcheddn && *matcheddn ) {
			printf( _("Matched DN: %s\n"), matcheddn );
		}

		if( refs ) {
			int i;
			for( i=0; refs[i]; i++ ) {
				printf(_("Referral: %s\n"), refs[i] );
			}
		}
	}

	/* if we were told to be quiet, use the return value. */
	if ( !quiet ) {
		if ( code == LDAP_COMPARE_TRUE ) {
			printf(_("TRUE\n"));
		} else if ( code == LDAP_COMPARE_FALSE ) {
			printf(_("FALSE\n"));
		} else {
			printf(_("UNDEFINED\n"));
		}
	}

	if ( ctrls ) {
		tool_print_ctrls( ld, ctrls );
		ldap_controls_free( ctrls );
	}

	ber_memfree( text );
	ber_memfree( matcheddn );
	ber_memvfree( (void **) refs );

	return( code );
}

